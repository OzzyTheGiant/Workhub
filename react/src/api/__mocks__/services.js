const services = { // all services must successHandler(an object literal that will update the state
	login:jest.fn().mockImplementation(({data, successHandler, errorHandler}) => {
		if (data.username === "OzzyTheGiant" && data.password === "1234567") {
			successHandler({
				isLoggedIn:true,
				currentModule:{
					name:"documents",
					listType:"clients",
					filePath:null,
					textFileData:null,
					parentId:null
				}
			});
		} else {
			errorHandler({response:{status:401}});
		}
	}),

    logout:jest.fn().mockImplementation(({successHandler, errorHandler}) => successHandler({isLoggedIn:false})),

    getClients:jest.fn().mockImplementation(({successHandler, errorHandler}) => {
		successHandler({
			clients:{
				"100000":{id:"100000", clientName: "Ozzy Perez"},
				"100001":{id:"100001", clientName: "Alondra Perez"}
			}
		});
	}),

    getProjects:jest.fn().mockImplementation(({successHandler, errorHandler}) => {
		successHandler({
			projects: {
				1:{
					category:{description:"Tax Return", id:1},
					dateCreated:"2018-04-26T19:58:20.000+0000",
					dateDue:"2017-04-15",
					id:1,
					name:"2017 Form 1040",
					client:"100000"
				}
			}
		});
	}),
	
	getDocuments:jest.fn().mockImplementation(({successHandler, errorHandler}) => {
		successHandler({
			documents:{
				"AAAAAAAAAAA":{
					accessLevel:"PUBLIC",
					category:{description:"Workpapers", id:3},
					description:"file2",
					fileType:"PDF",
					id:"AAAAAAAAAAA",
					year:2017,
					project:1
				},
				"ASDFHOWEOUH":{
					accessLevel:"PUBLIC",
					category:{description:"Notes", id:4},
					description:"notes",
					fileType:"TXT",
					id:"ASDFHOWEOUH",
					year:2017,
					project:1,
					client:"100000"
				},
				"BBBBBBBBBBB":{
					accessLevel:"PUBLIC",
					category:{description:"Workpapers", id:3},
					description:"file3",
					fileType:"EXCEL",
					id:"BBBBBBBBBBB",
					year:2017,
					project:1,
					client:"100000"
				}
			}
		});
    }),
    
	openDocument:jest.fn().mockImplementation(({data:docID, successHandler, errorHandler}) => {
		let filePath;
		switch(docID) {
			case "AAAAAAAAAAA": successHandler("/[100000] Ozzy Perez/[Tax Return] 2017 Form 1040/[AAAAAAAAAAA][Work Papers][2017] file2.pdf"); break;
			case "ASDFHOWEOUH": 
				filePath = "/[100000] Ozzy Perez/[Tax Return] 2017 Form 1040/[ASDFHOWEOUH][Notes][2017] notes.txt";
				services.downloadFile({data:filePath, successHandler, errorHandler});
				successHandler((prevState, props) => {
					return { currentModule:{...prevState.currentModule, filePath} };
				}); break;
			case "BBBBBBBBBBB":
				filePath = "/[100000] Ozzy Perez/[Tax Return] 2017 Form 1040/[BBBBBBBBBBB][Work Papers][2017] file3.xlsx"
				services.downloadFile({data:filePath, successHandler, errorHandler});
				successHandler((prevState, props) => {
					return { currentModule:{...prevState.currentModule, filePath} };
				}); break;
			default: successHandler(null); break;
		}
	}),

    downloadFile:jest.fn().mockImplementation(({data:filePath, successHandler, errorHandler}) => { 
		if (filePath && filePath.split(".").pop() === "txt") successHandler((prevState, props) => {
			return { currentModule:{...prevState.currentModule, textFileData:"Testing!"} };
		});
	}),

	getDocumentHistory:jest.fn().mockImplementation(({data:docId, successHandler, errorHandler}) => {
		successHandler({
			documentHistory:[
				{
					actionType:"OPEN",
					id:1,
					employee:{
						firstName:"John",
						middleName:null,
						lastName:"Doe",
						role:"STAFF",
						id:2
					}
				}
			]
		});
	})
};

export default services;