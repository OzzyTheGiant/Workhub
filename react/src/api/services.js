import ajax from 'api/ajax-manager';
import qs from 'qs';
import downloadDocument from 'utilities/document-downloader';

const services = { // all services must return an object literal that will update the state
	login:({data, successHandler, errorHandler}) => ajax.post("login", qs.stringify(data), {
		withCredentials:false,
		headers: {'Content-Type':'application/x-www-form-urlencoded'}
	}).then(response => successHandler({
		isLoggedIn:true,
		currentModule:{
			name:"documents",
			listType:"clients",
			filePath:null,
			textFileData:null,
			parentId:null
		}
	})).catch(error => errorHandler(error)),

    logout:({successHandler, errorHandler}) => ajax.get("logout", {
        withCredentials:false
    }).then(response => successHandler({isLoggedIn:false})).catch(error => errorHandler(error)),

    getClients:({successHandler, errorHandler}) => ajax.get("clients")
    .then(response => { 
		successHandler({clients:response.data.reduce(arrayToObjectConverter, {})});
	}).catch(error => errorHandler(error)),

    getProjects:({successHandler, errorHandler}) => ajax.get("projects")
    .then(response => { 
		successHandler({projects:response.data.reduce(arrayToObjectConverter, {})});
	}).catch(error => errorHandler(error)),
	
	getDocuments:({successHandler, errorHandler}) => ajax.get("documents")
    .then(response => { 
		successHandler({documents:response.data.reduce(arrayToObjectConverter, {})});
	}).catch(error => errorHandler(error)),
	
    // getProjectsByClientID:(clientID, successHandler, errorHandler) => ajax.get(`clients/${clientID}/projects`)
    // .then(response => successHandler(response.data)).catch(error => errorHandler(error)),

    // getDocumentsByProjectID:(projectID, successHandler, errorHandler) => ajax.get(`projects/${projectID}/documents`)
    // .then(response => successHandler(response.data)).catch(error => errorHandler(error)),
    
    openDocument:({data:docID, successHandler, errorHandler}) => ajax.get(`documents/${docID}/open`, {responseType:'text'}) // data is document ID
    .then(response => {
		let filePath = response.data;
		if (filePath.split(".").pop() !== ("pdf")) services.downloadFile({data:filePath, successHandler, errorHandler});
		else successHandler((prevState, props) => {
			return { currentModule:{...prevState.currentModule, filePath} };
		});
	}).catch(error => errorHandler(error)),

    downloadFile:({data:filePath, successHandler, errorHandler}) => ajax.get(`document-files${filePath}`, {
		responseType:filePath.split(".").pop() === "txt" ? 'text' : 'blob'
	}).then(response => {
        let url = response.config.url.split("/");
        if (response.config.responseType === 'text') {
			successHandler((prevState, props) => {
				return { currentModule:{...prevState.currentModule, textFileData:response.data} };
			});
		} else {
			successHandler((prevState, props) => {
				return { currentModule:{...prevState.currentModule, filePath} };
			});
			downloadDocument(url[url.length - 1], response.data, response.headers["content-type"])
		};
    }).catch(error => errorHandler(error)),

    getDocumentHistory:({data:docID, successHandler, errorHandler}) => ajax.get(`documents/${docID}/history`)
    .then(response => successHandler({documentHistory:response.data})).catch(error => errorHandler(error))
};

function arrayToObjectConverter (accumulator, entity, i, entities) {
	/* convert from array to object with id keys */
	accumulator[entity.id] = entities[i];
	return accumulator;
}

export default services;

