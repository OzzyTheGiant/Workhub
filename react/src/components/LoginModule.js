import React from "react";
import PropTypes from "prop-types";
import Notification from "components/Notification";

class LoginModule extends React.Component {
	onLoginFormSubmit = (e) => {
		if (e) e.preventDefault();
		this.props.serviceCaller("login", {
			username:this.props.currentModule.username,
			password:this.props.currentModule.password
		});
	}

	render() {
		return (
			<div id="LoginModule">
				<form action="/login" method="post" onSubmit={this.onLoginFormSubmit}>
					<img src="/workhub-icon-blue.svg" alt="Workhub Logo"/>
					<h1>Log in to your account</h1>
					<input type="text" name="username" placeholder="Username" onChange={(e) => this.props.updateModuleState({username: e.target.value})}/>
					<input type="password" name="password" placeholder="Password" onChange={(e) => this.props.updateModuleState({password: e.target.value})}/>
					<input type="submit" className="color-button button-ok" value="Log In"/>
				</form>
				<Notification type={this.props.notification.type ? "error" : ""} text={this.props.notification.text}/>
			</div>
		);
	}
}

LoginModule.propTypes = {
	currentModule:PropTypes.objectOf(PropTypes.string).isRequired,
	notification:PropTypes.objectOf(PropTypes.string),
	serviceCaller:PropTypes.func.isRequired,
	updateModuleState:PropTypes.func.isRequired
};

export default LoginModule;