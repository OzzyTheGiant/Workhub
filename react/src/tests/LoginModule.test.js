import React from 'react';
import { configure, mount, shallow } from 'enzyme';
import Adapter from 'enzyme-adapter-react-16';
import toJson from 'enzyme-to-json';
import LoginModule from 'components/LoginModule';
jest.mock('api/services');
import services from 'api/services';

describe('LoginModule', () => {
	let appState = {}, errorHandler = jest.fn(() => appState.notification = {type:"error", text:"Login Error: Username or password is incorrect!"});
	/* ==== MOCKs ==== */
	const updateState = jest.fn().mockImplementation((newState) => {
		appState = {...appState, ...newState};
	});
	const updateModuleState = jest.fn().mockImplementation((newState) => {
		/* update app's state, stored in appState variable */
		appState.currentModule = {...appState.currentModule, ...newState};
	});
	const serviceCaller = jest.fn().mockImplementation(function(service, postData) {
		/* when testing componentDidMount(), instead of calling App's method, it calls this function
			* which will return fake data from mock services module. Instead of setting props directly on currentWrapper,
			* due to circular reference issues and componentWrapper being null at the time of creating the successHandler function,
			* a brand new component and wrapper will have to be instantiated but with previous component's props */
		switch(service) {
			case "login": services.login({
				data:{username:appState.currentModule.username, password:appState.currentModule.password},
				successHandler:updateState,
				errorHandler
			}); break;
		} 
	});

    /* ==== COMPONENT ==== */
	const initComponent = () => {
		return (
			<LoginModule
			notification={{}}
			serviceCaller={serviceCaller}
			updateModuleState={updateModuleState} 
			currentModule={{
				name:"login",
				username:null,
				password:null
			}}/>
		);
	};
    let componentWrapper = null, loginModule = null;

	beforeAll(() => {
		configure({adapter:new Adapter()})
	});

	beforeEach(() => {
		loginModule = initComponent()
		componentWrapper = shallow(loginModule)
		appState = {...loginModule.props, isLoggedIn:false};
	});

	afterEach(() => {
		appState.username = appState.password = componentWrapper =  null;
	});

	it("Should render correctly", () => {
		const wrapper = mount(initComponent());
		let jsonComponent = toJson(wrapper);
		expect(jsonComponent.type).toBe('LoginModule');
		expect(jsonComponent.children[0].type).toBe('div');
		expect(jsonComponent).toMatchSnapshot();
	});

	it("Should update state when username or password is typed in", () => {
		componentWrapper.find('input[type="text"]').simulate('change', {target:{value:"OzzyTheGiant"}});
		componentWrapper.find('input[type="password"]').simulate('change', {target:{value:"1234567"}});
		componentWrapper.setProps(appState);
		expect(componentWrapper.instance().props.currentModule.username).toBe('OzzyTheGiant');
		expect(componentWrapper.instance().props.currentModule.password).toBe('1234567');
	});

	it("Should call login service with credentials and handlers on form submission", () => {
		componentWrapper.find('[type="text"]').simulate('change', {target:{value:"OzzyTheGiant"}});
       	componentWrapper.find('[type="password"]').simulate('change', {target:{value:"1234567"}});
		componentWrapper.find('form').simulate('submit');
		expect(services.login).toBeCalledWith({data:{username:"OzzyTheGiant", password:"1234567"}, successHandler:updateState, errorHandler});
		expect(appState.isLoggedIn).toBe(true);
		expect(appState.currentModule).toEqual({
			name:"documents",
			listType:"clients",
			filePath:null,
			textFileData:null,
			parentId:null
		});
	});

	it('Should call onFailedLogin() and display Notification with error message if login failed', () => {
		componentWrapper.find('[type="text"]').simulate('change', {target:{value:"OzzyTheGiant"}});
        componentWrapper.find('[type="password"]').simulate('change', {target:{value:"1234"}});
		componentWrapper.find('form').simulate('submit');
		componentWrapper.setProps(appState);
		expect(componentWrapper.find('Notification').prop("text")).toBe("Login Error: Username or password is incorrect!");
		expect(componentWrapper.find('Notification').prop("type")).toBe("error");
	});
});

