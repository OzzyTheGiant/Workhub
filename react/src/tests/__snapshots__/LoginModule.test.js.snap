// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LoginModule Should render correctly 1`] = `
<LoginModule
  currentModule={
    Object {
      "name": "login",
      "password": null,
      "username": null,
    }
  }
  notification={Object {}}
  serviceCaller={[Function]}
  updateModuleState={[Function]}
>
  <div
    id="LoginModule"
  >
    <form
      action="/login"
      method="post"
      onSubmit={[Function]}
    >
      <img
        alt="Logo"
        src="https://placehold.it/50x50/00638E"
      />
      <h1>
        Log in to your account
      </h1>
      <input
        name="username"
        onChange={[Function]}
        placeholder="Username"
        type="text"
      />
      <input
        name="password"
        onChange={[Function]}
        placeholder="Password"
        type="password"
      />
      <input
        className="color-button button-ok"
        type="submit"
        value="Log In"
      />
    </form>
    <Notification
      type=""
    >
      <div
        className="Notification none"
      />
    </Notification>
  </div>
</LoginModule>
`;
